servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/LoadBalancerObserver.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Copyright © 2023 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.loadbalancer;
+
+import io.servicetalk.client.api.NoActiveHostException;
+import io.servicetalk.client.api.ServiceDiscovererEvent;
+
+import java.util.Collection;
+import javax.annotation.Nullable;
+
+/**
+ * An observer that provides visibility into a {@link io.servicetalk.client.api.LoadBalancer}
+ * @param <ResolvedAddress> the type of the resolved address.
+ */
+interface LoadBalancerObserver<ResolvedAddress> {
+
+    /**
+     * Get a {@link HostObserver}.
+     * @return a {@link HostObserver}.
+     */
+    HostObserver<ResolvedAddress> hostObserver();
+
+    /**
+     * Callback for when connection selection fails due to no hosts being available.
+     */
+    void noHostsAvailable();
+
+    /**
+     * Callback for monitoring the changes due to a service discovery update.
+     */
+    void serviceDiscoveryEvent(Collection<? extends ServiceDiscovererEvent<ResolvedAddress>> events,
+                               int oldHostSetSize, int newHostSetSize);
+
+    /**
+     * Callback for when connection selection fails due to all hosts being inactive.
+     */
+    void noActiveHostsAvailable(int hostSetSize, NoActiveHostException exn);
+
+    interface HostObserver<ResolvedAddress> {
```

Comment:
Should we also define selectivity factors here, regardless if the current implementation uses them?

Reply:
Can you elaborate? I think you're considering something like "score changed" or something like that but I want to be sure. If that is the case, I'd slightly prefer to just add them when we have an example since.

Reply:
Correct, i was thinking something like `selectionOddsChanged(host, odds)`
Defining odds in a meaningful consumable form, sounds interesting. Example, we could introduce rational groups (low, normal, high) and use them to categorize the hosts. 
Happy to do in a later revision.

Reply:
Yes, that will be interesting but let's add it later when we have data to feed it.